### PR TITLE
PIL-818, PIL-827, SUP-122: Fix/key based migration improvements

### PR DESCRIPTION
**Pull request by poocart**
Includes:
- Improvement to wait until all transactions are created before showing success screen.
- Fix for showing what's being transferred.
- Deduct ETH from transfer amount if not enough balance left to pay fees.
- Other small fixes.

These fixes should cover:
- https://linear.app/pillarproject/issue/PIL-818/key-wallet-doesnt-migrate-funds-to-smart-wallet-after-pressing-magic
- https://linear.app/pillarproject/issue/PIL-827/user-can-not-transfer-eth-from-kw-to-sw
- https://linear.app/pillarproject/issue/SUP-122/the-application-crashes-during-migration

### CODE REVIEW

**Review by graszka22 — COMMENTED** [review]
What happened here? :D Did I just send a negative amount of ETH and the app said that it's okay? 🤣

**Comment by poocart** [comment]
> What happened here? :D Did I just send a negative amount of ETH and the app said that it's okay? 🤣
> 
> 

LOL! checking 👀

**Comment by poocart** [comment]
> > What happened here? :D Did I just send a negative amount of ETH and the app said that it's okay? 🤣
> > 
> 
> LOL! checking 👀

Fixed! Thanks for spotting.

**Review by graszka22 — APPROVED** [review]
LGTM 👍  you can consider adding tests for that, since more testing is our goal now and this function is so long and has ifs, calculations...

**Comment by poocart** [comment]
> LGTM 👍 you can consider adding tests for that, since more testing is our goal now and this function is so long and has ifs, calculations...

Added, thanks 👍

**Review by graszka22 — APPROVED** [review]

File: src/actions/keyBasedAssetTransferActions.js [hunk 1]
```diff
@@ -239,19 +242,70 @@ export const calculateKeyBasedAssetsToTransferTransactionGasAction = () => {
     const { history: { gasInfo } } = getState();
     const gasPrice = getGasPriceWei(gasInfo);
 
-    const keyBasedAssetsToTransferUpdated = await Promise.all(
+    let keyBasedAssetsToTransferUpdated = await Promise.all(
```
Comment (IAmKio) [inline src/actions/keyBasedAssetTransferActions.js:245]: Just wondering what happens if one of these fails?

File: src/actions/keyBasedAssetTransferActions.js [hunk 2]
```diff
@@ -239,19 +242,70 @@ export const calculateKeyBasedAssetsToTransferTransactionGasAction = () => {
...
 
+    /**
+     * ETH transfer amount adjustment below covers case if ETH is being transferred and sum of ETH amount
+     * that is being sent plus calculated ETH fees for token transfer is more than ETH balance that is left
+     * after ETH and token transfer, i.e. all ETH balance is sent and 0 left which is nt enough for other transfers
+     *
+     * in these case we want to subtract all transfer fees in ETH from ETH amount that is being sent that
+     * sum of ETH balance that is left after transfer plus ETH that is being transferred is lower than ETH fees
+     * needed to cover all transfer transactions
+     */
```
Comment (IAmKio) [inline src/actions/keyBasedAssetTransferActions.js:271]: Thanks for explaining this 👍

File: src/utils/__tests__/transactions.test.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      const calculatedEthTransactionAmountAfterFees = calculateETHTransactionAmountAfterFee(
+        ethTransactionAmount,
+        balances,
+        totalFeeInEth,
+      );
+      expect(calculatedEthTransactionAmountAfterFees).toStrictEqual(ethTransactionAmount);
+      expect(calculatedEthTransactionAmountAfterFees.isPositive()).toBeTruthy();
+    });
+  });
+});
```
Comment (IAmKio) [inline src/utils/__tests__/transactions.test.js:81]: Incredible, beautiful, fantastic! ❤️

**Review by IAmKio — APPROVED** [review]
Approving - thanks for the tests and comment blocks, it really helps 👍 Just one question but it's not a showstopper!

File: src/actions/keyBasedAssetTransferActions.js (hunk 1 above)
Comment (poocart) [reply src/actions/keyBasedAssetTransferActions.js:245]: It shouldn't, apart from basic mapping the only call we make is `calculateGasEstimate` and if it fails it returns default gas limit of 500k.

Comment (IAmKio) [reply src/actions/keyBasedAssetTransferActions.js:245]: Thanks for the explanation 👍

**Pull request merged**